Userland/Libraries/LibJS/Runtime/AsyncFromSyncIteratorPrototype.cpp

```diff
@@ -0,0 +1,205 @@
+    // 2. IfAbruptRejectPromise(done, promiseCapability).
+    auto done = TRY_OR_REJECT(vm, promise_capability, iterator_complete(global_object, result));
+
+    // 3. Let value be IteratorValue(result).
+    // 4. IfAbruptRejectPromise(value, promiseCapability).
+    auto value = TRY_OR_REJECT(vm, promise_capability, iterator_value(global_object, result));
+
+    // 5. Let valueWrapper be PromiseResolve(%Promise%, value).
+    // 6. IfAbruptRejectPromise(valueWrapper, promiseCapability).
+    auto value_wrapper = TRY_OR_REJECT(vm, promise_capability, promise_resolve(global_object, *global_object.promise_constructor(), value));
+
+    // 7. Let unwrap be a new Abstract Closure with parameters (value) that captures done and performs the following steps when called:
+    auto unwrap = [done](VM& vm, GlobalObject& global_object) -> ThrowCompletionOr<Value> {
+        // a. Return ! CreateIterResultObject(value, done).
+        return create_iterator_result_object(global_object, vm.argument(0), done);
+    };
+
+    // 8. Let onFulfilled be ! CreateBuiltinFunction(unwrap, 1, "", « »).
+    auto onFulfilled = NativeFunction::create(global_object, "", move(unwrap));
+    // 9. NOTE: onFulfilled is used when processing the "value" property of an IteratorResult object in order to wait for its value if it is a promise and re-package the result in a new "unwrapped" IteratorResult object.
+    VERIFY(is<Promise>(value_wrapper));
+    auto* value_wrapper_promise = static_cast<Promise*>(value_wrapper);
+
+    // 10. Perform ! PerformPromiseThen(valueWrapper, onFulfilled, undefined, promiseCapability).
+    value_wrapper_promise->perform_then(move(onFulfilled), js_undefined(), promise_capability);
+
+    // 11. Return promiseCapability.[[Promise]].
+    return promise_capability.promise;
+}
+
+// 27.1.4.2.1 %AsyncFromSyncIteratorPrototype%.next ( [ value ] ), https://tc39.es/ecma262/#sec-%asyncfromsynciteratorprototype%.next
+JS_DEFINE_NATIVE_FUNCTION(AsyncFromSyncIteratorPrototype::next)
+{
+    // 1. Let O be the this value.
+    // 2. Assert: O is an Object that has a [[SyncIteratorRecord]] internal slot.
+    auto* this_object = TRY(typed_this_object(global_object));
+
+    // 3. Let promiseCapability be ! NewPromiseCapability(%Promise%).
+    auto promise_capability = TRY(new_promise_capability(global_object, global_object.promise_constructor()));
+
+    // 4. Let syncIteratorRecord be O.[[SyncIteratorRecord]].
+    auto& sync_iterator_record = this_object->sync_iterator_record();
+    // 5. If value is present, then
+    //     a. Let result be IteratorNext(syncIteratorRecord, value).
+    // 6. Else,
+    //     a. Let result be IteratorNext(syncIteratorRecord).
+    // 7. IfAbruptRejectPromise(result, promiseCapability).
+    auto* result = TRY_OR_REJECT(vm, promise_capability,
+        (vm.argument_count() > 0 ? iterator_next(sync_iterator_record, vm.argument(0))
+                                 : iterator_next(sync_iterator_record)));
```

Comment:
I wrapped all this into a ternary. The code is a little bit further from the spec text but does remove the weird initialization of a throw completion and the extra if.